src/hotspot/share/gc/g1/g1CollectedHeap.cpp

```diff
@@ -3367,15 +3367,31 @@ void G1CollectedHeap::fill_with_dummy_object(HeapWord* start, HeapWord* end, boo
   region->fill_with_dummy_object(start, pointer_delta(end, start), zap);
 }
 
-void G1CollectedHeap::start_codecache_marking_cycle_if_inactive() {
+void G1CollectedHeap::start_codecache_marking_cycle_if_inactive(bool full_gc) {
+  // We can reach here with an active code cache marking cycle either because
+  // the previous G1 concurrent marking cycle
+  //
+  // (1) was aborted and a full gc is about to begin or
+  // (2) it was undone because the heap occupancy was below the threshold after
+  //     the initiating young gc (see G1ConcurrentMark::post_concurrent_undo_start())
+  //
+  // Especially in case (2) it is important that the active codecache cycle was not
+  // finished when executing the G1 undo operation. If it was (repeatedly) finished
+  // then nmethods with frames in continuation StackChunks could appear as not
+  // on stack because they were not marked as on stack in the undone concurrent marking.
```

Comment:
```suggestion
  // Especially in case (2) it is important to note the active codecache cycle did not
  // finish (maybe "complete, i.e. did not mark through all live objects,"?) when executing the G1 concurrent undo operation. If it (repeatedly) did not finish/complete
  // then nmethods with frames in continuation StackChunks could appear as not
  // on stack because they were not marked as on stack in the undone concurrent markings.
```